### PR TITLE
Fix package dependency issues

### PR DESCRIPTION
**Pull request by ChristopherGS**
**Update: For students checking this PR, the fixes have now been applied to the commit history. Please update your local copy from the current master branch**

Fix for the `Unit Testing a Production ML Model` section

### CODE REVIEW

File: packages/gradient_boosting_model/tests/test_pipeline.py [hunk 1]
```diff
@@ -3,41 +3,6 @@
...
-    # Given
-    X_train, X_test, y_train, y_test = pipeline_inputs
-    assert config.model_config.drop_features in X_train.columns
-
-    # When
-    # We use the scikit-learn Pipeline private method `_fit` which is called
-    # by the `fit` method, since this allows us to access the transformed
-    # dataframe. For other models we could use the `transform` method, but
-    # the GradientBoostingRegressor does not have a `transform` method.
-    X_transformed, _ = pipeline.price_pipe._fit(X_train, y_train)
```
Comment (ChristopherGS) [inline packages/gradient_boosting_model/tests/test_pipeline.py]: @solegalli previously I was able to use the `_fit` method to access the transformed dataframe of the pipeline. This method seems to have been updated (possibly here: https://github.com/scikit-learn/scikit-learn/commit/88ce8cd77e7aae1726364d92690025f0027c844c) and is throwing an error. 

Any idea how to access the transformed dataframe for the `GradientBoostingRegressor` (since it doesn't have a `transform` method?)

File: exercise_notebooks/docker_exercise/requirements.txt [hunk 2]
```diff
@@ -1,2 +1,3 @@
 flask>=1.1.1,<1.2.0
+markupsafe==2.0.1  # https://github.com/aws/aws-sam-cli/issues/3661
```
Comment (ChristopherGS) [inline exercise_notebooks/docker_exercise/requirements.txt:2]: Key Section 5 fix

File: packages/gradient_boosting_model/requirements.txt [hunk 3]
```diff
@@ -1,13 +1,14 @@
 # ML requirements
-numpy>=1.18.1,<1.19.0
-scikit-learn>=0.22.1,<0.23.0
-pandas>=0.25.3,<0.26.0
-feature_engine>=0.3.1,<0.4.0
-joblib>=0.14.1,<0.15.0
+numpy>=1.20.0,<1.21.0
+pandas>=1.3.5,<1.4.0
```
Comment (ChristopherGS) [inline packages/gradient_boosting_model/requirements.txt:3]: Key section 4 fix 1

File: packages/gradient_boosting_model/test_requirements.txt [hunk 4]
```diff
@@ -1,11 +1,11 @@
 -r requirements.txt
 
 # testing requirements
-pytest>=5.3.2,<6.0.0
+pytest>=6.2.3,<6.3.0
 
 # old model for testing purposes
 # source code: https://github.com/trainindata/deploying-machine-learning-models/tree/master/packages/regression_model
-tid-regression-model>=2.0.20,<2.1.0
+tid-regression-model==3.1.2
```
Comment (ChristopherGS) [inline packages/gradient_boosting_model/test_requirements.txt:8]: Key section 4 fix 2

File: exercise_notebooks/prometheus_exercise/requirements.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
 Flask>=1.1.1,<1.2.0
+markupsafe==2.0.1  # https://github.com/aws/aws-sam-cli/issues/3661
```
Comment (ChristopherGS) [inline exercise_notebooks/prometheus_exercise/requirements.txt:2]: key fix for section 9

File: packages/ml_api/requirements/requirements.txt [hunk 6]
```diff
@@ -1,12 +1,13 @@
 # ML Model
-tid-gradient-boosting-model>=0.1.18,<0.2.0
+tid-gradient-boosting-model>=0.3.0,<0.4.0
```
Comment (ChristopherGS) [inline packages/ml_api/requirements/requirements.txt:2]: Key fix for section 9 (2/2)

Note required new publishing of gradient boosting model (so sklearn versions are compatible)

File: packages/gradient_boosting_model/tests/test_pipeline.py (hunk 1 above)
Comment (solegalli) [reply packages/gradient_boosting_model/tests/test_pipeline.py]: Not straightforward. Here some potential solutions:

https://stackoverflow.com/questions/54332654/how-to-analyse-the-intermediate-steps-of-sklearn-pipeline
https://stackoverflow.com/questions/48743032/get-intermediate-data-state-in-scikit-learn-pipeline?

In short, I think you need to apply all the transformations one after the other, just up to the one before the model.

Comment (solegalli) [reply packages/gradient_boosting_model/tests/test_pipeline.py]: Ok, I asked the sklearn community through the email list and one of the developers told me to do this: 

Using slicing: model[:-1].transform(X)

Comment (ChristopherGSTest) [reply packages/gradient_boosting_model/tests/test_pipeline.py]: That worked perfectly :)

```
from gradient_boosting_model import pipeline
from gradient_boosting_model.config.core import config

from functools import reduce


def test_pipeline_drops_unnecessary_features(pipeline_inputs):
    # Given
    X_train, X_test, y_train, y_test = pipeline_inputs
    assert config.model_config.drop_features in X_train.columns
    pipeline.price_pipe.fit(X_train, y_train)

    # When
    # We access the transformed inputs with slicing
    transformed_inputs = pipeline.price_pipe[:-1].transform(X_train)

    # Then
    assert config.model_config.drop_features in X_train.columns
    assert config.model_config.drop_features not in transformed_inputs.columns


def test_pipeline_transforms_temporal_features(pipeline_inputs):
    # Given
    X_train, X_test, y_train, y_test = pipeline_inputs

    # When
    # We access the transformed inputs with slicing
    transformed_inputs = pipeline.price_pipe[:-1].transform(X_train)

    # Then
    assert (
        transformed_inputs.iloc[0]["YearRemodAdd"]
        == X_train.iloc[0]["YrSold"] - X_train.iloc[0]["YearRemodAdd"]
    )
```
